Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs

```diff
@@ -1,237 +1,142 @@
-        foreach (var direction in SharedDirectionExtensions.RandomDirections().ToArray())
-        {
-            var newPos = puddlePos.Offset(direction);
-            if (visitedTiles.Contains(newPos))
-                continue;
-
-            visitedTiles.Add(newPos);
 
-            if (CanExpand(newPos, mapGrid, out var uid))
-                nextToExpand.Add(new PuddlePlacer(newPos, (EntityUid) uid));
-        }
-    }
-
-    private bool CanExpand(Vector2i newPos, IMapGrid mapGrid,
-        [NotNullWhen(true)] out EntityUid? uid)
+    /// <summary>
+    /// Check a tile is valid for solution allocation.
+    /// </summary>
+    /// <param name="srcUid">Entity Uid of original puddle</param>
+    /// <param name="srcPuddle">PuddleComponent attached to srcUid</param>
+    /// <param name="pos">at which to check tile</param>
+    /// <param name="mapGrid">helper param needed to extract entities</param>
+    /// <param name="puddle">either found or newly created PuddleComponent.</param>
+    /// <returns>true if tile is empty or occupied by a non-overflowing puddle (or a puddle close to being overflowing)</returns>
+    private bool CheckTile(EntityUid srcUid, PuddleComponent srcPuddle, EntityCoordinates pos, IMapGrid mapGrid,
+        [NotNullWhen(true)] out PuddleComponent? puddle)
     {
-        if (!mapGrid.TryGetTileRef(newPos, out var tileRef)
+        if (!mapGrid.TryGetTileRef(pos, out var tileRef)
             || tileRef.Tile.IsEmpty)
         {
-            uid = null;
+            puddle = null;
             return false;
         }
 
-        foreach (var entity in mapGrid.GetAnchoredEntities(newPos))
+        var puddleCurrentVolume = _puddleSystem.CurrentVolume(srcUid, srcPuddle);
+
+        foreach (var entity in mapGrid.GetAnchoredEntities(pos))
         {
-            IPhysBody? physics = null;
-            PuddleComponent? existingPuddle = null;
+            // If this is valid puddle check if we spread to it.
+            if (TryComp(entity, out PuddleComponent? existingPuddle))
+            {
+                // If current puddle has more volume than current we skip that field
+                if (_puddleSystem.CurrentVolume(existingPuddle.Owner, existingPuddle) >= puddleCurrentVolume)
+                {
+                    puddle = null;
+                    return false;
+                }
+
+                puddle = existingPuddle;
+                return true;
+            }
 
-            // This is an invalid location
-            if (Resolve(entity, ref physics, false)
-                && (physics.CollisionLayer & (int) CollisionGroup.Impassable) != 0)
+            // if not puddle is this tile blocked by an object like wall or door
+            if (TryComp(entity, out PhysicsComponent? physComponent)
+                && physComponent.CanCollide
+                && ((physComponent.CollisionLayer + physComponent.CollisionMask) & (int) CollisionGroup.MobMask) != 0)
```

Comment:
collision masks and layers shouldn't be mixed in this way.

Reply:
Ok. Then how?